features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+
+// Present AN UI to let the user to optionally name the TOTP authenticator.
+
+const displayName = textInput.value;
+const { authenticatorID, secret } = await skygear.auth.mfa.createNewTOTP(displayName);
+
+// Present to the secret to the user and let the user
+// to add the secret to their TOTP application, such as Authy and Google Authenticator.
+
+// Present an UI to instruct the user to input the OTP.
+const otp = textInput.value;
+const { recoveryCodes } = await skygear.auth.mfa.activateTOTP(authenticatorID, otp);
+```
+
+```typescript
+// Present an UI to let the user to choose authenticator.
+// Suppose the user chose OOB.
+
+// Present an UI to instruct the user to input a phone number.
+const e164number = textInput.value;
+const { authenticatorID } = await skygear.auth.mfa.createNewOOB({ phone: e164number });
+
+// Present an UI to instruct the user to check SMS.
+
+// Present an UI to offer the user to trigger again in case the delivery failed.
+await triggerOOB(authenticatorID);
+
+// Present an UI to instruct the user to input the recevied code.
+const code = textInput.value;
+const { recoveryCodes } = await skygear.auth.mfa.activateOOB(authenticatorID, code);
+```
+
+```typescript
+// Check if recoveryCodes is present and present them to the user.
+// Ask the user to remember them and keep them safe.
+if (recoveryCodes) {
+  navigateToDisplayRecoveryCodeScreen(recoveryCodes);
+}
+```
+
+## Authentication
+
+```typescript
+try {
+  await skygear.auth.login("user@example.com", "password");
+} catch (e) {
+  // Re-raise other error.
+  if (!skygear.isMFARequiredError(e)) {
+    throw e;
+  }
+
+  let done = false;
+
+  const bearerToken = await getBearerToken();
+  if (bearerToken) {
+    try {
+      await skygear.auth.mfa.authenticateWithBearerToken(bearerToken);
+      done = true;
+    } catch (ee) {
+      if (!skygear.isMFARequiredError(ee)) {
+        throw ee;
+      }
+    }
+  }
+
+  if (done) {
+    return;
+  }
+
+  const authenticators = await skygear.auth.mfa.listAuthenticators();
+  // Present the authenticators to the user and let them choose
+  // which one they want to use.
+
+  // Suppose they chose OOB
+  await skygear.mfa.triggerOOB(authenticators[0].id);
+
+  // Present an UI to instruct the user to check SMS.
+  // Present an UI to offer the user to trigger resend.
+  // Present an UI to instruct the user to input the received code.
+  // Present an UI to offer the user to skip MFA for 30 days.
+
+  const code = textInput.value;
+  const { bearerToken } = await skygear.auth.mfa.authenticateWithOOB({ authenticatorID, code }, { requestBearerToken: true});
+
+  // Persist the bearer token
+  if (bearerToken) {
+    await storeBearerToken(bearerToken);
+  }
+}
+```
+
+The above example demonstrates the flow of authenitcation in details.
+The bearer token should be handled transparently by the SDK so the actual code written by the developer should not need to be aware of bearer token.
```

Comment:
The above flow seems to be the explicit flow using bearer token. We should also show a sample flow for transparent flow using bearer token.

Reply:
OK. I will add it.